### DISCUSSION TITLE
Complete VSIX Dev Reqs for VS2022?

### DISCUSSION THREAD

**Discussion by nurbles** [VsixCommunity/Community.VisualStudio.Toolkit#216] [Q&A] [answered]
I've just installed VS2022 on a fresh, clean Win10Pro to test it out with our solutions, which are all plain, native C (with a couple C++ modules). So I only installed the C/C++ Desktop package from the VS Installer to start. Then I followed [these instructions](https://www.vsixcookbook.com/getting-started/your-first-extension.html) and when I load the project I created in the first step, I get an error about the project targeting an older .NET version. Oh, I changed my project name to "xLearn" because the name used in the instructions matches one of the extensions loaded by the 'essentials' pack install by following the requirements from the document.

I closed the IDE and restarted, then loaded the solution again and this time it offered to retarget it for me, so I accepted. Now the IDE is reporting 38 errors (the first is `CS0246	The type or namespace name 'Community' could not be found (are you missing a using directive or an assembly reference?)	D:\code\xLearn\xLearn\xLearnPackage.cs	Line:1`)

This is the first time I've _ever_ looked at creating an extension, so I know nothing, but the tutorial doesn't mention that I should expect so many errors. Ergo, I'm confused about where to go next.

(For reference, I've been a C programmer for about 40 years on DOS original, then Windows and linux, too. So I have a _little_ experience... :-) )

**Comment by reduckted** [comment]
Sounds like your project might be using the wrong target framework. Each version of `Community.VisualStudio.Toolkit` requires a specific target Framework version which matches the version that Visual Studio uses.

|Package|Target Framework|
|--------|-----------------|
|`Community.VisualStudio.Toolkit.14`|4.6.0|
|`Community.VisualStudio.Toolkit.15`|4.6.0|
|`Community.VisualStudio.Toolkit.16`|4.7.2|
|`Community.VisualStudio.Toolkit.17`|4.8.0|

**Reply by nurbles** [reply]
I cannot find a string anywhere within the project directory in the form `Community.VisualStudio.Tookkit.##`.

I did find multiple references to these two things, though:
- "Community.VisualStudio.VSCT" Version="16.0.29.6"
- "Community.VisualStudio.Toolkit" Version="17.0.76.303"

.
Those names and version numbers appear in many formats: in PackageReference statements (in the .csproj file), as JSON sections, as parts of a path, in NuGet ImportProject lines, etc.  I'm not sure that's of any use and I don't know what the VSCT thing is that seems 'stuck"(?) at version 16.

Just for reference, I allowed VS2022 to perform the retargeting of the project. As a first time C# user, I would expect that to correctly adjust _all_ references necessary -- is that hoping for too much?  Does project retargeting also have some required manual steps?  If so, it would be nice it the VS2022 retargeting tool made mention of that.  Sorry, that's for Microsoft if they happen to be lurking.

**Comment by nurbles** [comment, marked as answer]
OK.  The comment about `Community.VisualStudio.Toolkit.17` sent me on a hunt and I finally stumbled over another requirement that was not mentioned in the instructions I referenced in my OP.  I needed to install the [.NET 4.8 Framework Developer's Pack](https://dotnet.microsoft.com/download/dotnet-framework/thank-you/net48-developer-pack-offline-installer) After I did that, the project built with no errors.

Frankly, I'm quite surprised that this wasn't automatically included by Visual Studio when I selected its "Visual Studio Extension Development" package.  Apparently, someone, somewhere assumed that I would already have this installed, I guess.

**Comment by madskristensen** [comment]
@nurbles I've opened an internal bug for the extensibility team to take a look at missing components in the extensibility workload

**Reply by wade0016** [reply]
I looked at that internal bug, and have included my response below:

The 4.8 workload is not required for the extensibility projects included with the Microsoft.VisualStudio.Workload.VisualStudioExtension workload. The Vsix Community project templates are marking its [project templates as requiring 4.8](https://github.com/VsixCommunity/ExtensibilityTemplatePack/blob/master/src/2022/ProjectTemplates/EmptyVsixProject/VSIXProject.csproj#:~:text=%24if%24%20%28%24targetframeworkversion%24%20%3e%204.8%2c%24endif%24). If the projects truly require 4.8 functionality, then the vsixmanifest can include that as a prerequisite:
`<Prerequisite Id="Microsoft.Net.Component.4.8.TargetingPack" Version="[17.0,18.0)" DisplayName=".NET Framework 4.8 targeting pack" />`
Otherwise, I would recommend reverting the switch to 4.8.

It is also possible to let the user decide during project creation which version of the .NET Framework should be targeted by using the [RequiredFrameworkVersion](https://docs.microsoft.com/en-us/visualstudio/extensibility/requiredframeworkversion-element-visual-studio-templates?view=vs-2022) element in the vstemplate file:
`<RequiredFrameworkVersion>4.7.2</RequiredFrameworkVersion>`
